diff --git a/client/messages.d.ts b/client/messages.d.ts
index 14f3730e5b406fcfeffe22e6735a1e109f496202..1efefbbfe996f098f45f346ac094426931acf078 100644
--- a/client/messages.d.ts
+++ b/client/messages.d.ts
@@ -113,7 +113,7 @@ export interface SendMessageParams {
     /** A list of message formatting entities. When provided, the parseMode is ignored. */
     formattingEntities?: Api.TypeMessageEntity[];
     /** Should the link preview be shown? */
-    linkPreview?: boolean;
+    linkPreview?: boolean | { showAboveText: boolean};
     /** Sends a message with a file attached (e.g. a photo, video, audio or document). The message may be empty. */
     file?: FileLike | FileLike[];
     /** Optional JPEG thumbnail (for documents). Telegram will ignore this parameter unless you pass a .jpg file!<br/>
@@ -181,7 +181,7 @@ export interface EditMessageParams {
     /** A list of message formatting entities. When provided, the parseMode is ignored. */
     formattingEntities?: Api.TypeMessageEntity[];
     /** Should the link preview be shown? */
-    linkPreview?: boolean;
+    linkPreview?: boolean | { showAboveText: boolean};
     /** The file object that should replace the existing media in the message. Does nothing if entity was a Message */
     file?: FileLike;
     /** Whether to send the given file as a document or not. */
diff --git a/client/messages.js b/client/messages.js
index 78899794ddf8f71d7769bfd19227ed8670ea8548..b0f40c0ee6d0350e36a2814d576704ba1eb76953 100644
--- a/client/messages.js
+++ b/client/messages.js
@@ -461,6 +461,7 @@ entity,
             noforwards: noforwards,
             commentTo: commentTo,
             topMsgId: topMsgId,
+            invertMedia: typeof linkPreview === 'object' ? linkPreview.showAboveText : false,
         });
     }
     entity = await client.getInputEntity(entity);
@@ -508,6 +509,7 @@ entity,
             entities: message.entities,
             clearDraft: clearDraft,
             noWebpage: !(message.media instanceof tl_1.Api.MessageMediaWebPage),
+            invertMedia: typeof linkPreview === 'object' ? linkPreview.showAboveText : false,
             scheduleDate: schedule,
             noforwards: noforwards,
         });
@@ -525,6 +527,7 @@ entity,
             message: message.toString(),
             entities: formattingEntities,
             noWebpage: !linkPreview,
+            invertMedia: typeof linkPreview === 'object' ? linkPreview.showAboveText : false,
             replyTo: replyObject,
             clearDraft: clearDraft,
             silent: silent,
@@ -658,6 +661,7 @@ async function editMessage(client, entity, { message, text, parseMode, formattin
         id,
         message: text,
         noWebpage: !linkPreview,
+        invertMedia: typeof linkPreview === 'object' ? linkPreview.showAboveText : false,
         entities,
         media: inputMedia,
         replyMarkup: markup,
diff --git a/client/uploads.d.ts b/client/uploads.d.ts
index ff7af2d3be9b32162b5e1630d4b16ab0c7547649..f4768c29ea555638a7ebce7405bed56dba5c43e7 100644
--- a/client/uploads.d.ts
+++ b/client/uploads.d.ts
@@ -109,6 +109,10 @@ export interface SendFileInterface {
      * Used for threads to reply to a specific thread
      */
     topMsgId?: number | Api.Message;
+    /**
+     * Show media (usually link preview) on the top of message text
+     */
+    invertMedia?: boolean;
 }
 interface FileToMediaInterface {
     file: FileLike;
diff --git a/client/uploads.js b/client/uploads.js
index 9392c42342a97ee253be8295c1a086b690673dca..3721a18b92568236faca12837b3fb074760525ec 100644
--- a/client/uploads.js
+++ b/client/uploads.js
@@ -316,7 +316,7 @@ async function _fileToMedia(client, { file, forceDocument, fileSize, progressCal
 }
 exports._fileToMedia = _fileToMedia;
 /** @hidden */
-async function _sendAlbum(client, entity, { file, caption, forceDocument = false, fileSize, clearDraft = false, progressCallback, replyTo, attributes, thumb, parseMode, voiceNote = false, videoNote = false, silent, supportsStreaming = false, scheduleDate, workers = 1, noforwards, commentTo, topMsgId, }) {
+async function _sendAlbum(client, entity, { file, caption, forceDocument = false, fileSize, clearDraft = false, progressCallback, replyTo, attributes, thumb, parseMode, voiceNote = false, videoNote = false, silent, supportsStreaming = false, scheduleDate, workers = 1, noforwards, commentTo, topMsgId, invertMedia = false, }) {
     entity = await client.getInputEntity(entity);
     let files = [];
     if (!Array.isArray(file)) {
@@ -408,13 +408,14 @@ async function _sendAlbum(client, entity, { file, caption, forceDocument = false
         scheduleDate: scheduleDate,
         clearDraft: clearDraft,
         noforwards: noforwards,
+        invertMedia: invertMedia,
     }));
     const randomIds = albumFiles.map((m) => m.randomId);
     return client._getResponseMessage(randomIds, result, entity);
 }
 exports._sendAlbum = _sendAlbum;
 /** @hidden */
-async function sendFile(client, entity, { file, caption, forceDocument = false, fileSize, clearDraft = false, progressCallback, replyTo, attributes, thumb, parseMode, formattingEntities, voiceNote = false, videoNote = false, buttons, silent, supportsStreaming = false, scheduleDate, workers = 1, noforwards, commentTo, topMsgId, }) {
+async function sendFile(client, entity, { file, caption, forceDocument = false, fileSize, clearDraft = false, progressCallback, replyTo, attributes, thumb, parseMode, formattingEntities, voiceNote = false, videoNote = false, buttons, silent, supportsStreaming = false, scheduleDate, workers = 1, noforwards, commentTo, topMsgId, invertMedia = false, }) {
     if (!file) {
         throw new Error("You need to specify a file");
     }
@@ -444,6 +445,7 @@ async function sendFile(client, entity, { file, caption, forceDocument = false,
             forceDocument: forceDocument,
             noforwards: noforwards,
             topMsgId: topMsgId,
+            invertMedia: invertMedia,
         });
     }
     if (Array.isArray(caption)) {
@@ -491,6 +493,7 @@ async function sendFile(client, entity, { file, caption, forceDocument = false,
         scheduleDate: scheduleDate,
         clearDraft: clearDraft,
         noforwards: noforwards,
+        invertMedia: invertMedia,
     });
     const result = await client.invoke(request);
     return client._getResponseMessage(request, result, entity);
diff --git a/client/telegramBaseClient.d.ts b/client/telegramBaseClient.d.ts
index 0319ed02fd4b19d4c63bb24012d1efe96d787ff9..92fcb53b827bd94054f1b3d6101ea4aa89a4a37d 100644
--- a/client/telegramBaseClient.d.ts
+++ b/client/telegramBaseClient.d.ts
@@ -62,6 +62,8 @@ export interface TelegramClientParams {
     autoReconnect?: boolean;
     /** does nothing for now */
     sequentialUpdates?: boolean;
+    /** Whether to receive and dispatch updates. Defaults to true. */
+    receiveUpdates?: boolean;
     /** The threshold below which the library should automatically sleep on flood wait and slow mode wait errors (inclusive).<br/>
      *  For instance, if a FloodWaitError for 17s occurs and floodSleepThreshold is 20s, the library will sleep automatically.<br/>
      *  If the error was for 21s, it would raise FloodWaitError instead. defaults to 60 sec.*/
diff --git a/client/telegramBaseClient.js b/client/telegramBaseClient.js
index 661f96ca2731060c18d5b4303414771dd67ce42e..90af7984c4a0fc26e304fb09ba77b6c0159c42e0 100644
--- a/client/telegramBaseClient.js
+++ b/client/telegramBaseClient.js
@@ -37,6 +37,7 @@ const clientParamsDefault = {
     downloadRetries: 5,
     autoReconnect: true,
     sequentialUpdates: false,
+    receiveUpdates: true,
     floodSleepThreshold: 60,
     deviceModel: "",
     systemVersion: "",
@@ -120,6 +121,7 @@ class TelegramBaseClient {
         this._selfInputPeer = undefined;
         this.useWSS = clientParams.useWSS;
         this._securityChecks = !!clientParams.securityChecks;
+        this._receiveUpdates = clientParams.receiveUpdates !== false;
         if (this.useWSS && this._proxy) {
             throw new Error("Cannot use SSL with proxies. You need to disable the useWSS client param in TelegramClient");
         }
diff --git a/client/TelegramClient.js b/client/TelegramClient.js
index 227d74d83df6e745cc9faf6c0a1f425b1e571923..3debb5353f6e64886ead265890e2517f8eaa2383 100644
--- a/client/TelegramClient.js
+++ b/client/TelegramClient.js
@@ -1067,7 +1067,7 @@ class TelegramClient extends telegramBaseClient_1.TelegramBaseClient {
                 autoReconnect: this._autoReconnect,
                 connectTimeout: this._timeout,
                 authKeyCallback: this._authKeyCallback.bind(this),
-                updateCallback: updates_1._handleUpdate.bind(this),
+                updateCallback: this._receiveUpdates ? updates_1._handleUpdate.bind(this) : undefined,
                 isMainSender: true,
                 client: this,
                 securityChecks: this._securityChecks,
@@ -1086,7 +1086,7 @@ class TelegramClient extends telegramBaseClient_1.TelegramBaseClient {
             testServers: this.testServers,
         });
         if (!(await this._sender.connect(connection, false))) {
-            if (!this._loopStarted) {
+            if (this._receiveUpdates && !this._loopStarted) {
                 (0, updates_1._updateLoop)(this);
                 this._loopStarted = true;
             }
@@ -1100,7 +1100,7 @@ class TelegramClient extends telegramBaseClient_1.TelegramBaseClient {
             layer: AllTLObjects_1.LAYER,
             query: this._initRequest,
         }));
-        if (!this._loopStarted) {
+        if (this._receiveUpdates && !this._loopStarted) {
             (0, updates_1._updateLoop)(this);
             this._loopStarted = true;
         }
diff --git a/client/auth.js b/client/auth.js
index 166074a7bf3b3524f7eae8aa6fb05fae3b0bba8c..2c295561de2963a76d62b5c7f2483b99a63e3a92 100644
--- a/client/auth.js
+++ b/client/auth.js
@@ -48,7 +48,12 @@ exports.start = start;
 /** @hidden */
 async function checkAuthorization(client) {
     try {
-        await client.invoke(new tl_1.Api.updates.GetState());
+        if (client._receiveUpdates === false) {
+            await client.invoke(new tl_1.Api.users.GetUsers({ id: [new tl_1.Api.InputUserSelf()] }));
+        }
+        else {
+            await client.invoke(new tl_1.Api.updates.GetState());
+        }
         return true;
     }
     catch (e) {
diff --git a/client/users.js b/client/users.js
index 437132d67b623f262c3634a94a9c7a5673907df2..0d4c401754de3154499c3fff994f2cc64794fd0e 100644
--- a/client/users.js
+++ b/client/users.js
@@ -128,7 +128,12 @@ exports.isBot = isBot;
 /** @hidden */
 async function isUserAuthorized(client) {
     try {
-        await client.invoke(new tl_1.Api.updates.GetState());
+        if (client._receiveUpdates === false) {
+            await client.invoke(new tl_1.Api.users.GetUsers({ id: [new tl_1.Api.InputUserSelf()] }));
+        }
+        else {
+            await client.invoke(new tl_1.Api.updates.GetState());
+        }
         return true;
     }
     catch (e) {
